src/Build/Logging/TerminalLogger/TerminalLogger.cs

```diff
@@ -704,65 +732,10 @@ private void ProjectFinished(object sender, ProjectFinishedEventArgs e)
-                            // * the output path relative to the initial working directory, if it is under it
-                            // * the output path relative to the source root, if it is under it
-
-                            // full path fallback
-                            var projectDisplayPathSpan = outputPathSpan;
-                            var workingDirectorySpan = _initialWorkingDirectory.AsSpan();
-                            // under working dir case
-                            if (outputPathSpan.StartsWith(workingDirectorySpan, FileUtilities.PathComparison))
-                            {
-                                if (outputPathSpan.Length > workingDirectorySpan.Length
-                                    && (outputPathSpan[workingDirectorySpan.Length] == Path.DirectorySeparatorChar
-                                        || outputPathSpan[workingDirectorySpan.Length] == Path.AltDirectorySeparatorChar))
-                                {
-                                    projectDisplayPathSpan = outputPathSpan.Slice(workingDirectorySpan.Length + 1);
-                                }
-                            }
-                            // under source root case
-                            else if (project.SourceRoot is { Span: var sourceRootSpan } )
-                            {
-                                var relativePathFromWorkingDirToSourceRoot = Path.GetRelativePath(workingDirectorySpan.ToString(), sourceRootSpan.ToString()).AsSpan();
-                                if (outputPathSpan.StartsWith(sourceRootSpan, FileUtilities.PathComparison))
-                                {
-                                    if (outputPathSpan.Length > sourceRootSpan.Length
-                                        // offsets are -1 here compared to above for ***reasons***
-                                        && (outputPathSpan[sourceRootSpan.Length - 1] == Path.DirectorySeparatorChar
-                                            || outputPathSpan[sourceRootSpan.Length - 1] == Path.AltDirectorySeparatorChar))
-                                    {
-                                        projectDisplayPathSpan = Path.Combine(relativePathFromWorkingDirToSourceRoot.ToString(), outputPathSpan.Slice(sourceRootSpan.Length).ToString()).AsSpan();
-                                    }
-                                }
-                            }
-
-                            Terminal.WriteLine(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("ProjectFinished_OutputPath", CreateLink(uri, projectDisplayPathSpan.ToString())));
+                            (var projectDisplayPath, var urlLink) = DetermineOutputPathToRender(outputPathSpan, _initialWorkingDirectory.AsMemory(), project.SourceRoot);
+                            Terminal.WriteLine(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("ProjectFinished_OutputPath", CreateLink(urlLink, projectDisplayPath.ToString())));
```

Comment:
This is all of the same logic, just extracted into stateless methods to prevent this method from getting too long and unwieldy. This should make more sense when I send the 'support multiple outputs' PR, where we reuse the helpers more cleanly.